src/twisted/web/test/test_http_headers.py

```diff
@@ -668,3 +668,36 @@ def test_copy(self):
         # Verify that the orignal does not have it
         self.assertEqual(h.getRawHeaders("test\u00E1"), ["foo\u2603", "bar"])
         self.assertEqual(h.getRawHeaders(b"test\xe1"), [b"foo\xe2\x98\x83", b"bar"])
+
+
+class MixedHeadersTests(TestCase):
+    """
+    Tests for L{Headers}, mixing L{bytes} and L{str} arguments for methods
+    where that is permitted.
+    """
+
+    def test_addRawHeader(self) -> None:
+        """
+        L{Headers.addRawHeader} accepts mixed L{str} and L{bytes}.
+        """
+        h = Headers()
+        h.addRawHeader(b"bytes", "str")
+        h.addRawHeader("str", b"bytes")
+
+        self.assertEqual(h.getRawHeaders(b"Bytes"), [b"str"])
+        self.assertEqual(h.getRawHeaders("Str"), ["bytes"])
+
+    def test_setRawHeaders(self) -> None:
+        """
+        L{Headers.setRawHeaders} accepts mixed L{str} and L{bytes}.
+        """
+        h = Headers()
+        h.setRawHeaders(b"bytes", [b"bytes"])
```

Comment:
Maybe hypothesis can help here... but I didnt' had time to check hypothesis data generator.

The idea, is that when we have bytes, we should have a value that is never a valid ascii or utf-8. In this way, it should detect bugs caused by accidental decoding attempts.  

```suggestion
        h.setRawHeaders(b'key-with-broken-utf8-\xe2\x98', [b'value-with-broken-utf8-\xe2\x98'])
```

but this is not a blocker, as may other tests in Twisted are written in this way.

It's just someting that maybe we can keep in mind and have in new tests.

Reply:
Same &mdash; the typechecker doesn't care.